File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution:
        # N-> len(nums)
        # K -> avg size of possible solution list inside "ans" list
        # TC: O(K*(2^N))
        # x -> No. of the possible combination
        # SC: (K*x)
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        ans=[]
        temp=[]
        nums.sort()
        
        def rec(target,ind):
                if(target==0):
                        ans.append(temp[:])
                        return
                
                for i in range (ind,len(nums)):
                        if(nums[i]>target):
                                return
                        if(i==ind or nums[i]!=nums[i-1] ):                                
                                temp.append(nums[i])
                                rec(target-nums[i],i+1)
                                temp.pop()
                        
        
        rec(target,0)
        return ans
        

File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
        vector<vector<int>>res;
        void bt(vector<int>&arr,vector<int>&temp,int target,int ind){
               if(target==0){res.push_back(temp);return;}
                for(int i=ind;i<arr.size();i++){
                        if(i>ind && arr[i]==arr[i-1])continue;
                        if(target-arr[i]<0 )break;
                        temp.push_back(arr[i]);
                        bt(arr,temp,target-arr[i],i+1);
                        temp.pop_back();
                }
        }
public:
    vector<vector<int>> combinationSum2(vector<int>& arr, int target) {
            sort(arr.begin(),arr.end());
            vector<int>temp;
            bt(arr,temp,target,0);
            return res;
    }
};